### PR TITLE
use NB_USER rather than getpass.getuser()

### LINKED ISSUE #129: Fails to start if user from getpass.getuser can't run it.
### Bug description
See https://phabricator.wikimedia.org/T302164

On jupyter-rsession-proxy 2 and up we are getting errors when trying to start rstudio because it is being run from the user from `getpass.getuser()`. Jupyterhub adds a `USER` variable with the username from the authenticator, but that user is not a linux user in our case. All pods use the same linux user to keep access to shared storage. Since the user does not exist we get an error in the form 
```
ERROR Could not find details for server user 'Chicocvenancio'; LOGGED FROM: virtual rstudio::core::ProgramStatus rstudio::server::Options::read(int, char* const*, std::ostream&) src/cpp/server/ServerOptions.cpp:299
```

#### Expected behaviour
Grab the user default from `NB_USER` instead of `USER`.

### PR DESCRIPTION
**Pull request by vivian-rook**
Sorry for the open/close PRs this gets us unblocked in a way that doesn't use my repository.

Resolves #129

### CODE REVIEW

**Comment by manics** [comment]
Is this extension designed to only work with jupyter/docker-stacks images where `NB_USER` is guaranteed to be defined, or should it also work for more general cases?

**Comment by ryanlovett** [comment]
@manics It should work in the more general case. So either the authenticator needs to pass in something useful in @vivian-rook's deployment, or the username needs to be configurable here.

File: jupyter_rsession_proxy/__init__.py [hunk 1]
```diff
@@ -45,7 +45,7 @@ def rewrite_auth(response, request):
 
 def setup_rserver():
     def _get_env(port):
-        return dict(USER=getpass.getuser())
+        return dict(USER=os.environ.get('NB_USER'))
```
Comment (yuvipanda) [inline jupyter_rsession_proxy/__init__.py:48]: I think we should go through the following hierarchy:

1. If `JUPYTER_USER` is set, use that. That's the logged-in username, if using a JupyterHub
2. If `NB_USER` is set, use that. This defaults to `jovyan` I think?
3. Use `getpass.getuser()` if neither of these are set.

**Review by yuvipanda — CHANGES_REQUESTED** [review]
Thanks for working on this!

File: jupyter_rsession_proxy/__init__.py (hunk 1 above)
Comment (chicocvenancio) [reply jupyter_rsession_proxy/__init__.py:48]: Is `JUPYTER_USER` guaranteed to exist as a system user? I only see `JUPYTERHUB_USER` in PAWS, and that is not a system user in the container. 

The args to `--server-user` and `--user-identity` need to be a system user, as I understand it.

Comment (vivian-rook) [reply jupyter_rsession_proxy/__init__.py:48]: Should the rsession-proxy use the above order, but have additional logic to verify that it is also a system user?

Comment (yuvipanda) [reply jupyter_rsession_proxy/__init__.py:48]: sorry, I meant `JUPYTERHUB_USER` not `JUPYTER_USER`.

@vivian-rook I don't think it matters for rsession-proxy that it is a system user. Is that right, @ryanlovett?

Comment (chicocvenancio) [reply jupyter_rsession_proxy/__init__.py:48]: @yuvipanda I think the problem with #129 is that `Chicocvenancio` not a system user.

Comment (yuvipanda) [reply jupyter_rsession_proxy/__init__.py:48]: ah goddamit, apparently i left my reading comprehension skills at home. Sorry!

In that case, can we try:

1. `pwd.getpwuid(os.getuid())[0]`, which should return a username that's in the `pwd` database. 
2. If this fails, then use `NB_USER`.

How does that sound?

Comment (vivian-rook) [reply jupyter_rsession_proxy/__init__.py]: Something like this?

**Review by yuvipanda — APPROVED** [review]
This LGTM. @ryanlovett looks ok to you? 

Thanks a lot, @vivian-rook!

**Comment by ryanlovett** [comment]
@yuvipanda The only suggestion I have is to have a fallback in the call to retrieve NB_USER, e.g. `user = os.environ.get('NB_USER', getpass.getuser())`, so that if _everything_ fails then we're back to the previous behavior. What do you think?

File: jupyter_rsession_proxy/__init__.py [hunk 2]
```diff
@@ -43,9 +44,16 @@ def rewrite_auth(response, request):
             u = urlparse(request.uri)
             response.headers[header] = urlunparse(u._replace(path=u.path+v))
 
+def get_system_user():
+    try:
+        user = pwd.getpwuid(os.getuid())[0]
+    except:
+        user = os.environ.get('NB_USER')
```
[the remark this answers is not in the document]
Comment (yuvipanda) [reply jupyter_rsession_proxy/__init__.py:51]: @ryanlovett like this?

[the remark this answers is not in the document]
Comment (ryanlovett) [reply jupyter_rsession_proxy/__init__.py:51]: Yep, exactly.

**Comment by yuvipanda** [comment]
@vivian-rook if you can test this and it works, am happy to merge.

**Comment by chicocvenancio** [comment]
@yuvipanda, @vivian-rook works for me with local jupyterhub.

**Comment by vivian-rook** [comment]
@yuvipanda Works for me off the latest build in https://github.com/toolforge/paws/pull/173

**Comment by yuvipanda** [comment]
Thanks a lot, @vivian-rook @chicocvenancio and @ryanlovett!

**Pull request merged**